ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/TopkKudaf.java

```diff
@@ -16,41 +16,56 @@
 
 package io.confluent.ksql.function.udaf.topk;
 
-import io.confluent.ksql.function.KsqlAggregateFunction;
-import io.confluent.ksql.parser.tree.Expression;
-import io.confluent.ksql.util.ArrayUtil;
-import io.confluent.ksql.util.KsqlException;
 import org.apache.kafka.connect.data.Schema;
 import org.apache.kafka.streams.kstream.Merger;
 
+import java.lang.reflect.Array;
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
 
-public class TopkKudaf<T> extends KsqlAggregateFunction<T, T[]> {
+import io.confluent.ksql.function.KsqlAggregateFunction;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.util.ArrayUtil;
+import io.confluent.ksql.util.KsqlException;
+
+public class TopkKudaf<T extends Comparable<? super T>> extends KsqlAggregateFunction<T, T[]> {
+
   private final int topKSize;
-  private final Object[] tempTopKArray;
   private final Class<T> clazz;
   private final Schema returnType;
   private final List<Schema> argumentTypes;
+  private final Comparator<T> comparator;
 
+  @SuppressWarnings("unchecked")
   TopkKudaf(int argIndexInValue,
             int topKSize,
             Schema returnType,
             List<Schema> argumentTypes,
             Class<T> clazz) {
     super(argIndexInValue,
-        () -> (T[]) new Object[topKSize],
-        returnType,
-        argumentTypes
+        () -> (T[]) Array.newInstance(clazz, topKSize),
```

Comment:
This was deliberately changed to not use `Array.newInstance` as the maximum size of an Array via this method is 255

Reply:
Really? Wow, I'll fix them all up.

Reply:
yeah - you are right. I read the doc wrong

Reply:
To be fair, docs are misleading. They talk about a 255 dim limit, even on the version of the function that only takes a single length param, i.e. which only builds one dimensional arrays.